File: examples/swift/parsed_messages.swift
import Foundation

@main
struct ParsedMessagesRunner {
    static func main() {
        do {
            try run()
        } catch {
            fatalError("Swift parsed messages example failed: \(error)")
        }
    }

    static func run() throws {
        print("Parsed Message Stream Example")

        let bytes = decodeSample()
        let reader = try reader_from_bytes(bytes)

        let stream = try reader_messages(reader)
        let parsed = try message_stream_parsed(stream)

        var count = 0
        while true {
            let item = try parsed_message_stream_next(parsed)
            if !item.has_value {
                break
            }
            let value = item.value
            if value.is_json {
                print("Message \(count): parsed JSON \(value.json.toString())")
            } else {
                print("Message \(count): raw bytes size=\(value.bytes.len())")
            }
            count += 1
            if count >= 5 {
                break
            }
        }
    }
}

File: examples/c/parsed_messages.c
#include "common.h"
#include "mcapable_ffi.h"
#include <stdio.h>

static void print_last_error(void) {
  McapByteBuffer err = mcap_last_error();
  if (err.ptr && err.len) {
    fwrite(err.ptr, 1, err.len, stderr);
    fputc('\n', stderr);
    mcap_byte_buffer_free(err);
  }
}

int main(void) {
  SampleBuffer sample = sample_bytes();
  McapReader* reader = mcap_reader_from_bytes(sample.data, sample.len);
  free_sample_bytes(sample);
  if (!reader) {
    print_last_error();
    return 1;
  }

  McapMessageStream* stream = mcap_reader_messages(reader);
  if (!stream) {
    print_last_error();
    mcap_reader_free(reader);
    return 1;
  }

  McapParsedStream* parsed = mcap_message_stream_parsed(stream);
  if (!parsed) {
    print_last_error();
    return 1;
  }

  McapParsedValue value = {0};
  size_t count = 0;
  while (mcap_parsed_stream_next(parsed, &value)) {
    if (value.is_json) {
      printf("Message %zu: parsed JSON %.*s\n", count, (int)value.json.len, (char*)value.json.ptr);
    } else {
      printf("Message %zu: raw bytes size=%zu\n", count, value.bytes.len);
    }
    mcap_parsed_value_clear(&value);
    if (++count >= 5) {
      break;
    }
  }

  reader = mcap_parsed_stream_into_reader(parsed);
  if (!reader) {
    print_last_error();
    return 1;
  }
  mcap_reader_free(reader);
  return 0;
}
